File: BCO_classes/road.rb
=begin
Name: Bastien Gliech
Scrum: Lord
Peak: Ballmer's
=end


class Road

	# constructs new road object
	# Precondition: params hash is formatted correctly
	def initialize(params)
		@loc1 = params[:loc1]
		@loc2 = params[:loc2]
		@distance = params[:distance]
		@pheremone = params[:pheremone]
	end
	attr_reader :loc1, :loc2

	# compares two roads to see if they are equal
	# Returns: boolean describing if roads are equals
	def equals(road2)

		# uses location class equals method (TODO)
		road2.loc2.equals(loc1) && road2.loc1.equals(loc2)
	end

	# evaporates pheremone on this road, in accordance to 
	def evapPheremone(evapRate)
		pheremone *= evapRate
	end

	# returns String representation of this road 
	def inspect
		puts "Road: #{@loc1} <--> #{@loc2}, distance = #{@distance}"
	end

	# Precondition: location l is at one end of this road
	# Returns: Location at other end
	def otherLoc(location)

		# uses location class equals method (TODO)
		if location.equals(loc1)
			loc2
		else
			loc1
		end
	end
end

File: JACO/Road.java
/*
Name: Jarred Durant
Course: CIS 421 -- Artificial Intelligence
Instructor: Dr. Laura Grabowski
Assignment #4 -- Pathfinding in Middle-Earth with Ant Colony Optimization
*/

public class Road implements Comparable<Road> {
  public Location loc1;     // One endpoint of the road (order doesn't matter)
  public Location loc2;     // Other endpoint of the road (order doesn't matter)
  public int distance;      // Length of road
  public int roadQuality;   // Quality of road. Not used in this assignment.
  public int riskLevel;     // Risk level of road. Not used in this assignment.
  public double pheromone;  // Pheromone level on road.

  // Comparison method. 
  public int compareTo(Road other) { return loc1.compareTo(other.loc1); }

	// Precondition: roadInfo is formatted correctly
	// Constructs a new Road object
  public Road(String[] roadInfo) {
    for (Location l : ACOPathfinder.locations) {
      if (roadInfo[0].equals(l.name)) loc1 = l;
      if (roadInfo[1].equals(l.name)) loc2 = l;
    }
    distance = Integer.parseInt(roadInfo[2]);
    roadQuality = Integer.parseInt(roadInfo[3]);
    riskLevel = Integer.parseInt(roadInfo[4]);
    pheromone = 15.0;
  }

  // Evaporate the pheromone on this road, in accordance with evapRate
  public void evapPheromone(double evapRate) {
    pheromone *= evapRate;
  }

	// returns String representation of Road
  public String toString() {
    return loc1.toString() + " <--> " + loc2.toString() + ": Distance = " + distance;
  }

	// Precondition: Location l is at one end of this road
	// Returns: The Location at the other end of the road from l
  public Location otherLoc(Location l) {
    if (l.equals(loc2)) return loc1;
    else return loc2;
  }
}
